roles/kubernetes/kubeadm/tasks/main.yml

```diff
@@ -29,19 +29,22 @@
   delegate_to: "{{ groups['kube_control_plane'][0] }}"
   run_once: true
 
-- name: Calculate kubeadm CA cert hash
-  shell: |
-    set -o pipefail && openssl x509 -pubkey -in {{ kube_cert_dir }}/ca.crt | \
-    openssl {% if 'RSA' in kube_asymmetric_encryption_algorithm %}rsa{% elif 'ECDSA' in kube_asymmetric_encryption_algorithm %}ec{% else %}rsa{% endif %} -pubin -outform der 2>/dev/null | \
-    openssl dgst -sha256 -hex | sed 's/^.* //'
-  args:
-    executable: /bin/bash
-  register: kubeadm_ca_hash
+- name: Fetch CA certificate from control plane node
+  slurp:
+    src: "{{ kube_cert_dir }}/ca.crt"
+  register: ca_cert_content
   when:
     - kubeadm_ca_stat.stat is defined
     - kubeadm_ca_stat.stat.exists
   delegate_to: "{{ groups['kube_control_plane'][0] }}"
   run_once: true
+
+- name: Calculate kubeadm CA cert hash
+  set_fact:
+    kubeadm_ca_hash: "{{ (ca_cert_content.content | b64decode | community.crypto.x509_certificate_info).public_key_fingerprints.sha256 }}"
+  when: ca_cert_content.content | length > 0
+  delegate_to: localhost
+  run_once: true
```

Comment:
This is pretty good, great work !

Just a thought though: do we need to use set fact ? I think it should be possible to simply have that expression directly in the jinja template in kubeadm-client.conf.j2, wdyt ? (adjusting the conditionals).

The only downside I could see is that it would compute it for each host, but I don't think the difference is noticeable.

Reply:
directly placing the expression in the `kubeadm-client.conf.j2` template is indeed more concise.